Provide a Not Found handler

akrabat opened the pull request:
Implement `ProblemDetailsNotFoundHandler` so that an API can send back a problem details response if the request accepts XML or JSON.

Typical usage in Expressive is to add it into your pipeline immediate before the 
default `NotFoundHandler`:

```
$app->pipe(\Zend\ProblemDetails\ProblemDetailsNotFoundHandler::class);
$app->pipe(NotFoundHandler::class);
```

akrabat committed:
Implement ProblemDetailsNotFoundHandler

This handler returns a problem details Response object with a 404 status
code set along with the correctly generated problem details.

akrabat committed:
Add tests for ProblemDetailsNotFoundHandler

akrabat committed:
Add docs for ProblemDetailsNotFoundHandler

weierophinney commented on src/ProblemDetailsNotFoundHandler.php:
```diff
@@ -0,0 +1,60 @@
+<?php
```
This file needs the licence docblock.

akrabat committed:
Fix PHPCS report

weierophinney commented on test/ProblemDetailsNotFoundHandlerTest.php:
```diff
@@ -0,0 +1,73 @@
+    {
+        $request = $this->prophesize(ServerRequestInterface::class);
+        $request->getMethod()->willReturn('POST');
+        $request->getHeaderLine('Accept')->willReturn('application/json');
```
We should likely demonstrate XML, too.

weierophinney commented on docs/book/not-found-handler.md:
```diff
@@ -0,0 +1,23 @@
+that it will accept either JSON or XML.
+
+
+To use this handler in Expressive add it into your pipeline (usually `pipeline.php`)
```
This paragraph has one extra newline both before and after it for some reason...

weierophinney commented on docs/book/not-found-handler.md:
````diff
@@ -0,0 +1,23 @@
+$app->pipe(NotFoundHandler::class);
+```
+
+
````
These extra newlines at the end may be removed.

weierophinney commented on docs/book/quick-start.md:
````diff
@@ -280,3 +280,22 @@ $app->post('/api/domain/transaction', [
+$app->pipe(\Zend\ProblemDetails\ProblemDetailsNotFoundHandler::class);
+$app->pipe(NotFoundHandler::class);
+```
+
````
Please remove the extra newlines at the end here.

weierophinney commented on docs/book/quick-start.md:
````diff
@@ -280,3 +280,22 @@ $app->post('/api/domain/transaction', [
+To use this handler in Expressive add it into your pipeline immediate before the 
+default `NotFoundHandler`:
+
+```
````
Add the language following the opening code block, please.

weierophinney commented on docs/book/not-found-handler.md:
````diff
@@ -0,0 +1,23 @@
+immediate before the default `NotFoundHandler`:
+
+
+```
````
Please add the language to this opening code block.

weierophinney reviewed (changes requested):
A few minor nitpicks, and a request to demonstrate XML payload generation. Otherwise, looks solid, Rob!

michalbundyra commented on src/ProblemDetailsNotFoundHandler.php:
```diff
@@ -0,0 +1,60 @@
+    private $responseFactory;
+
+    /**
+     * @param ProblemDetailsResponseFactory $responseFactory Factory to create a response to
```
Type in PHPDocs should be consistent with method declaration: `null|ProblemDetailsResponseFactory`

michalbundyra commented on src/ProblemDetailsNotFoundHandler.php:
```diff
@@ -0,0 +1,60 @@
+     */
+    public function __construct(ProblemDetailsResponseFactory $responseFactory = null)
+    {
+        $this->responseFactory = $responseFactory ?: new ProblemDetailsResponseFactory();
```
As we required PHP7.1 we can have here: 
```php
 $responseFactory ?? new ProblemDetailsResponseFactory();
```

Not sure if it is more efficient or it doesn't matter at all?

akrabat replied:
Makes no difference. `?:` is consistent with `ProblemDetailsMiddleware::__construct()`

weierophinney replied:
`??` is mostly useful if the variable referenced might not be defined; it adds an implicit `isset`. Since we know it's defined (it's in the constructor), `?:` makes sense here.

michalbundyra commented on src/ProblemDetailsNotFoundHandler.php:
```diff
@@ -0,0 +1,60 @@
+    /**
+     * Can the middleware act as an error handler?
+     *
+     * Returns a boolean false if negotiation fails.
```
Here we have description of return value, so it should be with tag `@return`.

akrabat replied:
Updated, though arguably could just be removed.

akrabat committed:
Tidy ProbleDetailsNotFoundHandler documentation

akrabat committed:
Add license and docblock updates to ProbleDetailsNotFoundHandler

akrabat committed:
Update ProbleDetailsNotFoundHandler tests to accept XML and JSON

Note that as `testResponseFactoryPassedInConstructorGeneratesTheReturnedResponse()`
returns the prophesied response, the `$expectedHeader` isn't required.

akrabat commented:
Updated as requested.

michalbundyra commented on src/ProblemDetailsNotFoundHandler.php:
```diff
@@ -0,0 +1,66 @@
+     * Creates and returns a 404 response.
+     *
+     * @param ServerRequestInterface $request Ignored.
+     * @param DelegateInterface $delegate Ignored.
```
Hm... missed before, sorry. These params description is wrong. These params are not ignored at all. We can remove that part of docs block to not duplicate with method signature.

weierophinney replied:
I can do this on merge.

One other thing I've discovered recently: return types in PHP are _covariant_, which means you can specify a return type _more specific_ than the one defined in the interface or parent class. That means we can add `: ResponseInterface` here safely to make it self-documenting. :smile:

michalbundyra commented on src/ProblemDetailsNotFoundHandler.php:
```diff
@@ -0,0 +1,66 @@
+    /**
+     * Can the middleware act as an error handler?
+     *
+     * @return boolean false if negotiation fails.
```
I think we are using `bool`, just for consistency.

michalbundyra commented on test/ProblemDetailsNotFoundHandlerTest.php:
```diff
@@ -0,0 +1,89 @@
+            "title" => "Not Found",
+            "type" => "https://httpstatus.es/404",
+            "status" => 404,
+            "detail" => "Cannot POST https://example.com/foo!",
```
Any reason to using double quotes above?

akrabat replied:
Side-effect of learning Swift :)

akrabat committed:
Remove unnecessary docblock elements

akrabat committed:
Add return type to ProblemDetailsNotFoundHandler::process()

akrabat committed:
Use single quotes in strings

akrabat commented:
Comments addressed.

michalbundyra commented on test/ProblemDetailsNotFoundHandlerTest.php:
```diff
@@ -0,0 +1,89 @@
+{
+    use ProblemDetailsAssertionsTrait;
+
+    public function acceptHeaders()
```
In the other test case (`ProblemDetailsNotFoundHandlerFactoryTest`) on all methods you have declared return type, but not in this test case. Can we add it please, for consistency? Thx 😄

akrabat committed:
Add return types to ProblemDetailsNotFoundHandlerTest

akrabat committed:
Ensure delegate is called when Accept is unacceptable

If the Accept header doesn't allow for sending a problem details format
response back in ProblemDetailsNotFoundHandler, then the handler should
do nothing and act as a no-op.

weierophinney commented:
Thanks, @Akrabat
